File: backend/src/modules/notificacion/notificacion.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '@common/config/database/prisma.service';
import { ListarNotificacionesDto } from './dto/listar-notificaciones.dto';

export interface NotificacionDetalle {
  notificacion: {
    id: string;
    canal: string;
    mensaje: string;
    leida: boolean;
    creadoEn: Date;
    pedidoId: string | null;
  };
  pedido:
    | {
        id: string;
        emailComprador: string;
        telefonoComprador: string;
        estado: string;
        totalCentavos: number;
        codigo: string;
        creadoEn: Date;
        confirmadoEn: Date | null;
        vencidoEn: Date;
      }
    | null;
}

@Injectable()
export class NotificacionService {
  constructor(private readonly prisma: PrismaService) {}

  async listar(filtros: ListarNotificacionesDto): Promise<
    Array<{
      id: string;
      canal: string;
      mensaje: string;
      leida: boolean;
      creadoEn: Date;
      pedidoId: string | null;
    }>
  > {
    const where: { leida?: boolean } = {};
    if (filtros.filtro === 'unread') {
      where.leida = false;
    }

    return this.prisma.notificacion.findMany({
      where,
      orderBy: { creadoEn: 'desc' },
    });
  }

  async obtenerDetalle(id: string): Promise<NotificacionDetalle> {
    const notificacion = await this.prisma.notificacion.findUnique({
      where: { id },
      include: { pedido: true },
    });

    if (!notificacion) {
      throw new NotFoundException('Notificación no encontrada');
    }

    const pedidoTransformado = notificacion.pedido
      ? {
          id: notificacion.pedido.id,
          emailComprador: notificacion.pedido.emailComprador,
          telefonoComprador: notificacion.pedido.telefonoComprador,
          estado: notificacion.pedido.estado,
          totalCentavos: Number(notificacion.pedido.totalCentavos),
          codigo: notificacion.pedido.codigo,
          creadoEn: notificacion.pedido.creadoEn,
          confirmadoEn: notificacion.pedido.confirmadoEn,
          vencidoEn: notificacion.pedido.vencidoEn,
        }
      : null;

    return {
      notificacion,
      pedido: pedidoTransformado,
    };
  }

  async marcarComoLeida(id: string): Promise<{
    id: string;
    leida: boolean;
    creadoEn: Date;
  }> {
    const notificacion = await this.prisma.notificacion.findUnique({
      where: { id },
    });

    if (!notificacion) {
      throw new NotFoundException('Notificación no encontrada');
    }

    if (notificacion.leida) {
      return notificacion;
    }

    return this.prisma.notificacion.update({
      where: { id },
      data: { leida: true },
      select: { id: true, leida: true, creadoEn: true },
    });
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/notificacion/NotificacionService.java
package com.sistemapedidos.modules.notificacion;

import com.sistemapedidos.common.error.NotFoundException;
import com.sistemapedidos.domain.entity.Notificacion;
import com.sistemapedidos.domain.entity.Pedido;
import com.sistemapedidos.domain.repository.NotificacionRepository;
import com.sistemapedidos.modules.notificacion.dto.ListarNotificacionesDto;
import com.sistemapedidos.modules.notificacion.dto.NotificacionDetalle;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class NotificacionService {

    private final NotificacionRepository notificacionRepository;

    @Transactional(readOnly = true)
    public List<NotificacionDetalle.NotificacionDto> listar(ListarNotificacionesDto filtros) {
        List<Notificacion> notificaciones = "unread".equals(filtros.filtro())
                ? notificacionRepository.findByLeidaFalseOrderByCreadoEnDesc()
                : notificacionRepository.findAllByOrderByCreadoEnDesc();
        return notificaciones.stream()
                .map(this::aDto)
                .toList();
    }

    @Transactional(readOnly = true)
    public NotificacionDetalle obtenerDetalle(String id) {
        Notificacion notificacion = notificacionRepository.findByIdConPedido(id)
                .orElseThrow(() -> new NotFoundException("Notificacion no encontrada"));

        Pedido pedido = notificacion.getPedido();
        NotificacionDetalle.PedidoResumido pedidoResumido = pedido != null
                ? new NotificacionDetalle.PedidoResumido(
                        pedido.getId(),
                        pedido.getEmailComprador(),
                        pedido.getTelefonoComprador(),
                        pedido.getEstado().name(),
                        pedido.getTotalCentavos(),
                        pedido.getCodigo(),
                        pedido.getCreadoEn(),
                        pedido.getConfirmadoEn(),
                        pedido.getVencidoEn())
                : null;

        return new NotificacionDetalle(aDto(notificacion), pedidoResumido);
    }

    @Transactional
    public Map<String, Object> marcarComoLeida(String id) {
        Notificacion notificacion = notificacionRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Notificacion no encontrada"));

        if (!notificacion.getLeida()) {
            notificacion.setLeida(true);
            notificacionRepository.save(notificacion);
        }

        return Map.of("id", notificacion.getId(), "leida", notificacion.getLeida(), "creadoEn", notificacion.getCreadoEn());
    }

    private NotificacionDetalle.NotificacionDto aDto(Notificacion notificacion) {
        return new NotificacionDetalle.NotificacionDto(
                notificacion.getId(),
                notificacion.getCanal().name(),
                notificacion.getMensaje(),
                notificacion.getLeida(),
                notificacion.getCreadoEn(),
                notificacion.getPedido() != null ? notificacion.getPedido().getId() : null);
    }
}
